sdk/core/System.ClientModel/src/ModelReaderWriter/ModelReaderWriterOptions.cs

```diff
@@ -33,4 +38,115 @@ public ModelReaderWriterOptions (string format)
+    /// </summary>
+    /// <param name="proxy"> The <see cref="IPersistableModel{T}"/> proxy that will be used to read or write the model. </param>
+    public void AddProxy<T>(IPersistableModel<T> proxy)
+    {
+        if (_proxies is null)
+            _proxies = new Dictionary<Type, object>();
+
+        _proxies.Add(typeof(T), proxy);
+    }
+
+    /// <summary>
+    /// Gets the model that is currently being proxied.
+    /// </summary>
+    /// <remarks>
+    /// Whenever a proxy is used for a given type, the model being proxied is set here.
+    /// The value does not need to be cleared after setting it will be overwritten on the next use.
+    /// </remarks>
+    public object? ProxiedModel { get; private set; }
+
+    /// <summary>
+    /// Gets the <see cref="IPersistableModel{T}"/> proxy for the specified <typeparamref name="T"/> model type.
+    /// </summary>
+    /// <param name="proxy"> The <see cref="IPersistableModel{T}"/> proxy if one exists. </param>
+    /// <returns> True if a proxy was found; otherwise, false. </returns>
+    public bool TryGetProxy<T>([NotNullWhen(true)] out IPersistableModel<T>? proxy)
+    {
+        if (_proxies is null || !_proxies.TryGetValue(typeof(T), out object? result))
+        {
+            proxy = default;
+            return false;
+        }
+
+        proxy = (IPersistableModel<T>)result;
+        return true;
+    }
+
+    /// <summary>
+    /// Gets the <see cref="IJsonModel{T}"/> proxy for the specified <typeparamref name="T"/> model type.
+    /// </summary>
+    /// <param name="proxy"> The <see cref="IJsonModel{T}"/> proxy if one exists. </param>
+    /// <returns> True if a proxy was found; otherwise, false. </returns>
+    public bool TryGetProxy<T>([NotNullWhen(true)] out IJsonModel<T>? proxy)
+    {
+        if (_proxies is null || !_proxies.TryGetValue(typeof(T), out object? result) || result is not IJsonModel<T> jsonResult)
+        {
+            proxy = default;
+            return false;
+        }
+
+        proxy = jsonResult;
+        return true;
+    }
+
+    internal bool TryGetProxy(Type modelType, [NotNullWhen(true)] out IJsonModel<object>? proxy)
+    {
+        if (_proxies is null || !_proxies.TryGetValue(modelType, out object? result) || result is not IJsonModel<object> jsonResult)
+        {
+            proxy = default;
+            return false;
+        }
+
+        proxy = jsonResult;
+        return true;
+    }
+
+    internal bool TryGetProxy(Type modelType, [NotNullWhen(true)] out IPersistableModel<object>? proxy)
+    {
+        if (_proxies is null || !_proxies.TryGetValue(modelType, out object? result))
+        {
+            proxy = default;
+            return false;
+        }
+
+        proxy = (IPersistableModel<object>)result;
+        return true;
+    }
+
+    /// <summary>
+    /// Gets the <see cref="IPersistableModel{T}"/> proxy for the specified <typeparamref name="T"/> model type.
```

Comment:
It would be good to add an explanation of what "resolve" means and how it works either here or in `<remarks>` in these docs, as this piece is a little non-standard for .NET patterns.